File: core/src/main/java/cloudspec/validator/AssertValidationResult.java
/*-
 * #%L
 * CloudSpec Core Library
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package cloudspec.validator;

import java.util.List;
import java.util.Optional;

public class AssertValidationResult {
    private final List<String> path;
    private final Boolean success;
    private final Optional<AssertValidationError> error;

    public AssertValidationResult(List<String> path, Boolean success) {
        this(path, success, null);
    }

    public AssertValidationResult(List<String> path, Boolean success, AssertValidationError error) {
        this.path = path;
        this.success = success;
        this.error = Optional.ofNullable(error);
    }

    public List<String> getPath() {
        return path;
    }

    public Boolean getSuccess() {
        return success;
    }

    public Optional<AssertValidationError> getError() {
        return error;
    }

    public Boolean isSuccess() {
        return success;
    }

    @Override
    public String toString() {
        return "AssertValidationResult{" +
                "path='" + path + '\'' +
                ", success=" + success +
                ", message='" + error + '\'' +
                '}';
    }
}

File: core/src/main/kotlin/cloudspec/validator/AssertValidationResult.kt
/*-
 * #%L
 * CloudSpec Core Library
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package cloudspec.validator

sealed class Path {
    fun toPathString(): String {
        return when (this) {
            is PropertyPath -> ".${this.name}"
            is NestedPropertyPath -> ".${this.name}"
            is KeyValuePropertyPath -> ".${this.name}[${this.key}]"
            is AssociationPath -> when {
                this.id.isNullOrEmpty() -> ">${this.name}"
                else -> ">${this.name}[${this.id}]"
            }
        }
    }
}

fun List<Path>.toPathString(): String {
    return when {
        this.isNotEmpty() -> "${this[0].toPathString()}${this.drop(1).toPathString()}"
        else -> ""
    }
}

data class PropertyPath(val name: String) : Path()
data class NestedPropertyPath(val name: String) : Path()
data class KeyValuePropertyPath(val name: String, val key: String) : Path()
data class AssociationPath(val name: String, val id: String? = null) : Path()

data class AssertValidationResult(
        val path: List<Path>,
        val success: Boolean,
        val error: AssertError? = null
)
